### PR TITLE
Add semantic segmentation

### LINKED ISSUE #1039: [FEAT] [CV] Add support for image detection with segmentation
**Is your feature request related to a problem? Please describe.**
 Add support for image detection with segmentation 

**Describe the solution you'd like**
We just need a more general IoU and area functions and a new data type that inherits from detection data with a new label and prediction validation

**Describe alternatives you've considered**
A custom data object

### PR DESCRIPTION
**Pull request by nirhutnik**
closes #1039 
(partially at least)

### What's done:
* Added SegmentationData
* Added Label properties
* Added MeanDice metric
* Made sure most checks work with this (details on which do not at the end of comment)

### What's left:
* Add tests for SegmentationData - [Will be completed in #1888]
* Add tests for MeanDice - [Will be completed in #1888]
* Add tests for changed checks (with working model) - [Will be completed in #1888]
* Change prediction format + Add prediction properties - V
* Update SegmentationData docstrings - V
* Update properties docstrings - V
* Update checks docstrings
* Update MeanDice docstrings - V
* Make new code more efficient - especially in MeanDice - [Will be completed in #1888]
* Image properties for segments - not like cropped photos (area = actual pixels. Return images with nan) - Will be done in #1889 

### Checks that currently don't work:
* ConfusionMatrixReport - Open issue with the rest:
* MeanAveragePrecisionReport
* MeanAverageRecallReport
* ModelErrorAnalysis
* RobustnessReport
* SimpleModelComparison (it shouldn't)

### CODE REVIEW

File: docs/source/user-guide/vision/tutorials/plot_custom_task_tutorial.py [hunk 1]
```diff
@@ -25,7 +25,7 @@
 4. `Implement Custom Metric <#implement-custom-metric>`__
 """
 
-#%%
+# %%
```
Comment (noamzbr) [inline docs/source/user-guide/vision/tutorials/plot_custom_task_tutorial.py:28]: why?

File: docs/source/user-guide/vision/tutorials/plot_custom_task_tutorial.py [hunk 2]
```diff
@@ -190,7 +190,7 @@ def batch_collate(batch):
     collate_fn=batch_collate
 )
 
-#%%
+# %%
```
Comment (noamzbr) [inline docs/source/user-guide/vision/tutorials/plot_custom_task_tutorial.py:193]: why?

File: docs/source/user-guide/vision/tutorials/plot_custom_task_tutorial.py [hunk 3]
```diff
@@ -203,9 +203,9 @@ def batch_collate(batch):
     axs[i].imshow(np.asarray(img))
     axs[i].set(xticklabels=[], yticklabels=[], xticks=[], yticks=[])
 
-fix
+fix.show()
 
-#%%
+# %%
```
Comment (noamzbr) [inline docs/source/user-guide/vision/tutorials/plot_custom_task_tutorial.py:208]: .

File: docs/source/user-guide/vision/tutorials/plot_custom_task_tutorial.py [hunk 4]
```diff
@@ -242,7 +242,8 @@ def batch_to_images(self, batch) -> Sequence[np.ndarray]:
         """Convert the batch to a list of images as (H, W, C) 3D numpy array per image."""
         return [tensor.numpy().transpose((1, 2, 0)) for tensor in batch[0]]
 
-#%%
+
+# %%
```
Comment (noamzbr) [inline docs/source/user-guide/vision/tutorials/plot_custom_task_tutorial.py:246]: .

File: docs/source/user-guide/vision/tutorials/plot_custom_task_tutorial.py [hunk 5]
```diff
@@ -253,8 +254,9 @@ def batch_to_images(self, batch) -> Sequence[np.ndarray]:
 test_task = MyCustomSegmentationData(test_data_loader)
 
 result = PropertyLabelCorrelationChange().run(train_task, test_task)
-result
+result.show()
 
+# %%
```
Comment (noamzbr) [inline docs/source/user-guide/vision/tutorials/plot_custom_task_tutorial.py:259]: .

File: docs/source/user-guide/vision/tutorials/plot_custom_task_tutorial.py [hunk 6]
```diff
@@ -291,11 +294,10 @@ def classes_in_labels(labels: List[torch.Tensor]) -> List[int]:
     {'name': 'Classes in Labels', 'method': classes_in_labels, 'output_type': 'class_id'}
 ]
 
-
 result = TrainTestLabelDrift(label_properties=label_properties).run(train_task, test_task)
-result
+result.show()
 
-#%%
+# %%
```
Comment (noamzbr) [inline docs/source/user-guide/vision/tutorials/plot_custom_task_tutorial.py:300]: .

File: docs/source/user-guide/vision/tutorials/plot_custom_task_tutorial.py (hunk 1 above)
Comment (nirhutnik) [reply docs/source/user-guide/vision/tutorials/plot_custom_task_tutorial.py:28]: because sometimes space is pressed accidentally.

File: docs/source/user-guide/vision/tutorials/plot_custom_task_tutorial.py (hunk 2 above)
Comment (nirhutnik) [reply docs/source/user-guide/vision/tutorials/plot_custom_task_tutorial.py:193]: Ok maybe something else happened.

File: docs/source/user-guide/vision/tutorials/plot_custom_task_tutorial.py (hunk 3 above)
Comment (nirhutnik) [reply docs/source/user-guide/vision/tutorials/plot_custom_task_tutorial.py:208]: I have no ideaaaaaaa

File: docs/source/user-guide/vision/tutorials/plot_custom_task_tutorial.py (hunk 4 above)
Comment (nirhutnik) [reply docs/source/user-guide/vision/tutorials/plot_custom_task_tutorial.py:246]: WHAT

File: docs/source/user-guide/vision/tutorials/plot_custom_task_tutorial.py (hunk 5 above)
Comment (nirhutnik) [reply docs/source/user-guide/vision/tutorials/plot_custom_task_tutorial.py:259]: WHAT IS HAPPENING

File: docs/source/user-guide/vision/tutorials/plot_custom_task_tutorial.py (hunk 6 above)
Comment (nirhutnik) [reply docs/source/user-guide/vision/tutorials/plot_custom_task_tutorial.py:300]: WHO DID THIS

File: deepchecks/vision/checks/train_test_validation/property_label_correlation_change.py [hunk 7]
```diff
@@ -136,7 +136,16 @@ def update(self, context: Context, batch: Batch, dataset_kind: DatasetKind):
                     class_id = int(label[0])
                     imgs += [cropped_img]
                     target += [dataset.label_id_to_name(class_id)]
-            property_type = PropertiesInputType.BBOXES
+            property_type = PropertiesInputType.PARTIAL_IMAGES
+        elif dataset.task_type == TaskType.SEMANTIC_SEGMENTATION:
+            for img, label in zip(batch.images, batch.labels):
+                for seg_label in np.unique(label):
+                    seg_location = label == seg_label
```
Comment (matanper) [inline deepchecks/vision/checks/train_test_validation/property_label_correlation_change.py:143]: why do you override label herer?

File: docs/source/user-guide/vision/tutorials/plot_custom_task_tutorial.py [hunk 8]
```diff
@@ -154,7 +154,7 @@ def load_or_download(cls, root: Path, train: bool) -> 'CocoSegmentDataset':
                     extract_root=str(root),
                     md5=md5
                 )
-            
+
```
Comment (matanper) [inline docs/source/user-guide/vision/tutorials/plot_custom_task_tutorial.py:157]: why

File: docs/source/user-guide/vision/tutorials/plot_custom_task_tutorial.py [hunk 9]
```diff
@@ -269,6 +271,7 @@ def batch_to_images(self, batch) -> Sequence[np.ndarray]:
 
 from deepchecks.vision.checks import TrainTestLabelDrift
 
+
```
Comment (matanper) [inline docs/source/user-guide/vision/tutorials/plot_custom_task_tutorial.py:274]: .

File: docs/source/user-guide/vision/tutorials/plot_custom_task_tutorial.py (hunk 8 above)
Comment (nirhutnik) [reply docs/source/user-guide/vision/tutorials/plot_custom_task_tutorial.py:157]: Because useless spaces are useless, of course.

File: deepchecks/vision/metrics_utils/semantic_segmentation_metrics.py [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+        y_pred, y = output
+
+        for i in range(len(y)):
+            for class_id in [int(x) for x in torch.unique(y[i])]:
+                y_pred_this_label = (y_pred[i].argmax(0) == class_id).numpy()
+                y_this_label = (y[i] == class_id).numpy()
+
+                tp = np.logical_and(y_pred_this_label, y_this_label).sum()
+                total_y_pred = y_pred_this_label.sum()
+                total_y = y_this_label.sum()
```
Comment (TheSolY) [inline deepchecks/vision/metrics_utils/semantic_segmentation_metrics.py:46]: Since the sum is on bool, I'd call it y_counts and y_pred_counts rather then "total", but not critical

**Review by TheSolY — CHANGES_REQUESTED** [review]
Overall good.
I've added the missing tests to https://github.com/deepchecks/deepchecks/issues/1888

File: deepchecks/vision/datasets/segmentation/segmentation_coco.py [hunk 11]
```diff
@@ -0,0 +1,375 @@
...
+    """
+
+    def batch_to_labels(self, batch):
+        """Extract from the batch only the labels by the given format in SegmentationData class."""
+        return batch[1]
+
+    def infer_on_batch(self, batch, model, device):
+        """Infer on a batch of images by the given format in SegmentationData class."""
+        normalized_batch = [F.normalize(img.unsqueeze(0).float()/255,
+                                        mean=(0.485, 0.456, 0.406), std=(0.229, 0.224, 0.225)) for img in batch[0]]
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/vision/datasets/segmentation/segmentation_coco.py:60]: https://pytorch.org/vision/0.12/auto_examples/plot_visualization_utils.html?highlight=visualization%20utils

File: deepchecks/vision/datasets/segmentation/segmentation_coco.py [hunk 12]
```diff
@@ -0,0 +1,375 @@
...
+
+        predictions = [model(img)["out"].squeeze(0).detach() for img in normalized_batch]
+        predictions = [torch.nn.functional.softmax(pred, dim=0) for pred in predictions]
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/vision/datasets/segmentation/segmentation_coco.py:63]: It's (C, H, W) where C is the classes.

File: deepchecks/vision/datasets/segmentation/segmentation_coco.py [hunk 13]
```diff
@@ -0,0 +1,375 @@
...
+        assert not all(l is None for l in labels), 'Did not find folder with labels or it was empty'
+
+        train_len = int(self.TRAIN_FRACTION * len(images))
+
+        if self.train is True:
+            self.images = images[0:train_len]
+            self.labels = labels[0:train_len]
+        else:
+            self.images = images[train_len:]
+            self.labels = labels[train_len:]
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/vision/datasets/segmentation/segmentation_coco.py:198]: 1. I don't train on this data
2. This is just as an example code to show and test the code
3. leakage is good for testing the checks :P
4. This is a straight-up copy-paste from our detection/coco.py

File: deepchecks/vision/datasets/segmentation/segmentation_coco.py [hunk 14]
```diff
@@ -0,0 +1,375 @@
...
+            root: str,
+            name: str,
+            train: bool = True,
+            transforms: t.Optional[t.Callable] = None,
+    ) -> None:
+        super().__init__(root, transforms=transforms)
+
+        self.train = train
+        self.root = Path(root).absolute()
+        self.images_dir = Path(root) / 'images' / name
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/vision/datasets/segmentation/segmentation_coco.py:169]: This is a straight-up copy-paste from our detection/coco.py. I think the "/" are an operator of Path. Anyway this passes the tests both for linux and windows

File: deepchecks/vision/metrics_utils/scorers.py [hunk 15]
```diff
@@ -82,7 +89,11 @@ def get_default_object_detection_scorers() -> t.Dict[str, Metric]:
     'fnr_micro': lambda: ObjectDetectionTpFpFn(evaluating_function='fnr', averaging_method='micro'),
     'fnr_weighted': lambda: ObjectDetectionTpFpFn(evaluating_function='fnr', averaging_method='weighted'),
     'average_precision_per_class': lambda: ObjectDetectionAveragePrecision(return_option='ap'),
-    'average_recall_per_class': lambda: ObjectDetectionAveragePrecision(return_option='ar')
+    'average_recall_per_class': lambda: ObjectDetectionAveragePrecision(return_option='ar'),
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/vision/metrics_utils/scorers.py:92]: Cause I first added a metric there and then removed. Will remove :)

File: deepchecks/vision/metrics_utils/semantic_segmentation_metrics.py [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+        self._evals = defaultdict(lambda: {'dice': 0, 'count': 0})
+
+    def update(self, output: Tuple[torch.Tensor, torch.Tensor]):
+        """Update metric with batch of samples."""
+        y_pred, y = output
+
+        for i in range(len(y)):
+            for class_id in [int(x) for x in torch.unique(y[i])]:
+                y_pred_this_label = (y_pred[i].argmax(0) == class_id).numpy()
+                y_this_label = (y[i] == class_id).numpy()
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/vision/metrics_utils/semantic_segmentation_metrics.py:42]: :+1:

File: deepchecks/vision/metrics_utils/semantic_segmentation_metrics.py (hunk 10 above)
Comment (nirhutnik) [reply deepchecks/vision/metrics_utils/semantic_segmentation_metrics.py:46]: :+1:

File: deepchecks/vision/checks/model_evaluation/confusion_matrix.py [hunk 17]
```diff
@@ -47,7 +47,7 @@ def filter_confusion_matrix(confusion_matrix: pd.DataFrame, number_of_categories
 class ConfusionMatrixReport(SingleDatasetCheck):
     """Calculate the confusion matrix of the model on the given dataset.
 
-    For object detection, each detected bounding box calculates the IoU for each label and then is that label class is
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/vision/checks/model_evaluation/confusion_matrix.py:50]: I just removed the "is" :P

File: deepchecks/vision/checks/train_test_validation/property_label_correlation_change.py [hunk 18]
```diff
@@ -136,7 +136,7 @@ def update(self, context: Context, batch: Batch, dataset_kind: DatasetKind):
                     class_id = int(label[0])
                     imgs += [cropped_img]
                     target += [dataset.label_id_to_name(class_id)]
-            property_type = PropertiesInputType.BBOXES
+            property_type = PropertiesInputType.PARTIAL_IMAGES
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/vision/checks/train_test_validation/property_label_correlation_change.py]: Too bad I changed it 😅

File: deepchecks/vision/segmentation_data.py [hunk 19]
```diff
@@ -0,0 +1,193 @@
...
+        ----------
+        batch : torch.Tensor
+            The batch of data.
+
+        Returns
+        -------
+        List[torch.Tensor]
+            List of torch.Tensors. Each tensor should be of same width and height of corresponding image, and its values
+            are the true class_ids of the corresponding pixels in that image.
+            Note that the tensor should 2D, as the number of channels on the original image are irrelevant to the class.
```
Comment (noamzbr) [inline deepchecks/vision/segmentation_data.py:48]: Not the clearest sentence (suggesting to improve it, but really not a must).
Maybe would be clearer to just state the shape (H, W) or something like that

File: deepchecks/vision/datasets/segmentation/segmentation_coco.py [hunk 20]
```diff
@@ -0,0 +1,375 @@
...
+        normalized_batch = [F.normalize(img.unsqueeze(0).float()/255,
+                                        mean=(0.485, 0.456, 0.406), std=(0.229, 0.224, 0.225)) for img in batch[0]]
+
+        predictions = [model(img)["out"].squeeze(0).detach() for img in normalized_batch]
+        predictions = [torch.nn.functional.softmax(pred, dim=0) for pred in predictions]
+
+        return predictions
+
+    def batch_to_images(self, batch) -> Sequence[np.ndarray]:
+        """Convert the batch to a list of images as (H, W, C) 3D numpy array per image."""
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/vision/datasets/segmentation/segmentation_coco.py:68]: :+1:

File: deepchecks/vision/segmentation_data.py [hunk 21]
```diff
@@ -0,0 +1,193 @@
...
+            The batch of data.
+        model : torch.nn.Module
+            The model to use for inference.
+        device : torch.device
+            The device to use for inference.
+
+        Returns
+        -------
+        Sequence[torch.Tensor]
+            The predictions of the model on the batch. The predictions should be in a sequence of length N containing
```
Comment (noamzbr) [inline deepchecks/vision/segmentation_data.py:99]: Think we should not that the values should be proba and sum to 1

File: requirements/dev-requirements.txt [hunk 22]
```diff
@@ -2,7 +2,7 @@ twine
 
 pylint==2.13.5
 pydocstyle 
-flake8 
+flake8==4.0.1
```
Comment (noamzbr) [inline requirements/dev-requirements.txt:5]: hmm? Must we have that exact version?

File: deepchecks/vision/segmentation_data.py (hunk 19 above)
Comment (nirhutnik) [reply deepchecks/vision/segmentation_data.py:48]: :+1:

File: deepchecks/vision/segmentation_data.py (hunk 21 above)
Comment (nirhutnik) [reply deepchecks/vision/segmentation_data.py:99]: :+1:

File: requirements/dev-requirements.txt (hunk 22 above)
Comment (nirhutnik) [reply requirements/dev-requirements.txt:5]: Bug in new version, I do what JKL tells me

File: deepchecks/vision/datasets/segmentation/segmentation_coco.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (TheSolY) [reply deepchecks/vision/datasets/segmentation/segmentation_coco.py:63]: I suggest that you mention this in the docstring, also to avoid future format inconsistency related bugs

**Review by TheSolY — APPROVED** [review]
LGTM 11/8/22

**Comment by nirhutnik** [comment]
Closes #1887

**Review by shir22 — APPROVED** [review]

**Review by ItayGabbay — APPROVED** [review]
